BlobUriBuilder preserve slashes

jaschrep-msft opened the pull request:
Addresses #24955
Addresses #30507

BlobUriBuilder strips leading and trailing slashes from blob names when constructing the Uri. Creating a client to a blob with a leading or trailing slash in blob name requires invoking the constructor directly (`BlobClient(Uri, Credential, BlobClientOptions`), which forces a developer to keep references to their credentials wherever they may need to create a blob client. While this and other modern Storage libraries make it very difficult to create a blob with a trailing slash in the name, it is still possible. Additionally, other libraries for interacting with Storage may be creating blobs with this behavior.  Clients should be capable of accessing these blobs in our recommended workflows.

Changes:
- BlobUriBuilder has an additional constructor which takes a flag on whether to preserve leading and trailing slashes on blob name.
- BlobClientOptions has a bool property `PreserveBlobNameSlashes` determining what flag to construct BlobUriBuilder with. Default is false.

jaschrep-msft committed:
BlobUriBuilder configurable for trailing slash preservation

jaschrep-msft committed:
exportapi

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[Azure.Storage.Blobs](https://apiview.dev/Assemblies/Review/a9af7ed90fe84a649de0b47c989eaf07)

jaschrep-msft committed:
null check

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobUriBuilder.cs:
```diff
@@ -34,6 +34,19 @@ public class BlobUriBuilder
+        /// used by <see cref="BlobContainerClient.GetBlobClient(string)"/> and similar
+        /// methods. Defaults to false.
+        /// </summary>
+        public static bool PreserveTrailingSlashDefault { get; set; }
```
@JoshLove-msft @tg-msft @kasobol-msft Looking for feedback regarding a static property. A static property is not the way we generally want to add these hooks. However, we do want a more centralized access point to change these defaults, as this class is used under the hood in many different places, and it seemed this behavior did not make sense as a "client option."

Do we have best practices for how to expose such a hook for a customer to change default behavior, instead of being forced to call new overloads for every `BlobContainerClient.GetBlobClient()` call in their entire code base? Is client options still the best answer?

jaschrep-msft replied:
Moved back to client options after testing concerns. Still happy to discuss whether this is really a "client option" but it is in clientoptions for now.

jaschrep-msft committed:
now preserves leading slash on command as well

jaschrep-msft committed:
exportapi

jaschrep-msft commented:
"Do Not Merge" added as this PR contains API changes and the upcoming release has already gone through API review. Must wait for next one.

jaschrep-msft committed:
static property => client option

static property decision already ran into some confusion designing
tests. reverted design decision to use client option. Can be discussed
in API review before release.

jaschrep-msft committed:
exportapi

jaschrep-msft committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into uribuilder-preserveslashes

jaschrep-msft committed:
fixes and validatino

jaschrep-msft committed:
property rename

jaschrep-msft committed:
remove suppressions used for testing

jaschrep-msft committed:
tests

jaschrep-msft committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into uribuilder-preserveslashes

jaschrep-msft committed:
changelog
